### ISSUE TITLE
Support parameters "deleteFilesOlderThan" or   "deleteFilesIfMoreThan"  and "filesToDeletePattern" in file target

### ISSUE THREAD

**Issue by MNF** [NLog/NLog#150]
I want to write each entry in a new file and be able to delete if number of files with the same prefix ( generally speaking - with the same name pattern defined as RegEx) exceed the limit (e.g. 100) or if they older than determine period(e.g 2 days).
The similar suggestions can be found at 
http://stackoverflow.com/questions/7833431/rolling-file-log-in-console-application-using-nlog

> Start a new log file on each run. So each time the application is started, a new log file should be created.
>  i want to keep the x latest runs.

and at

http://nlog-forum.1685105.n2.nabble.com/Re-New-features-in-NLog-please-read-and-make-comments-td1685288.htm

> Such a "deleteFilesOlderThan" option would be roughly what we'd like, 
> but it would have to do so according to the filename pattern...so you'd 
> probably have to transform ${shortdate} into an appropriate regex match, 
> and only delete matching filenames older than the specified date.  It is 
> possible that different loggers would need different values, or that 
> there might exist non-log files in the same directory. 

Another Related discussion
http://stackoverflow.com/questions/5565597/delete-log-files-after-x-days/6157576#6157576

**Comment by nzain** [comment]
This gets important, when multiple, independent instances of your application run in parallel (think of visual studio) but you want to keep a single log for each run. My scenario:

- Create a new log file for each run. Possible via `fileName="foo${cached:${date:format=yyyyMMdd_HHmmss}}.log"`
- Multiple instances can run in parallel (thus archiving doesn't work for me, second process would move log file of the first one)
- Do not fill up the hard drive over the years! Thats the missing feature.

A simple but dangerous solution would be something like
```
deleteOldFilesPattern="foo*.log"
deleteOldFilesAge="5 days"
```
Difficult to test for the nlog-user, easy to get wrong. In my case, multiple apps are logging to `%programdata%\companyname\` directory and I would have to double check my glob pattern to avoid deleting any other log files.

A sophisticated, complex solution would infer, if a file matches the dynamic filename. Phew. A restricted support for ${shortdate} or ${longdate} is certainly doable.

**Comment by snakefoot** [comment]
Have you tried this configuration, that performs automatic archive operation by using dated-filename and also supports cleanup when too many files:

```
    fileName="${logDirectory}/Log.${date:format=yyyy-MM-dd-HH}.log"
    archiveFileName="${logDirectory}/Log.{#}.log"
    archiveDateFormat="yyyy-MM-dd-HH"
    archiveNumbering="Date"
    archiveEvery="Year"
    maxArchiveFiles="60"
    keepFileOpen="true"
    concurrentWrites="true"
```
- **archiveFileName** - Using {#} allows the archive cleanup to generate proper file wildcard.
- **archiveDateFormat** - Must match the ${date:format=} of the fileName (So remember to correct both date-formats, if change is needed)
- **archiveNumbering=Date** - Configures the archive cleanup to support parsing of filenames as dates.
- **archiveEvery=Year** - Activates the archive cleanup, but also the archive file operation. Because the configured fileName automatically ensures the archive file operation, then we don't want any additional archive operations (Ex. avoiding generating extra empty files at midnight).
- **maxArchiveFiles** - How many archive files to keep around (your suggested deleteFilesIfMoreThan)

With NLog 4.5 (Still in BETA), then it will be a lot easier (As one just have to specify MaxArchiveFiles). See also https://github.com/NLog/NLog/pull/1993

**Comment by nzain** [comment]
@snakefoot thanks, this combination did the trick. The only issue left, was that the file changes every hour (your code) or every second (my code). The `keepFileOpen` option looks intriguing, but does not what you expect in this configuration. I added `${cached: ... }` to keep the filename for the application lifetime and this really works as intended:
- Every application will get its own single log file (unless two apps are started in the same second)
- Apps can run in parallel, keep their original log file and don't interfere with each others log
- Old logs are deleted
- Each run of the application has a single log file, except for Dezember 31. at midnight... where the log is split. Acceptable, will never happen for my customers.

My final solution for future reference and other readers:

```xml
    <target name="f" xsi:type="File"
            
                   fileName="logs/Log ${cached:${date:format=yyyy-MM-dd HH_mm_ss}}.log"
            archiveFileName="logs/Log {#}.log"
            archiveDateFormat="yyyy-MM-dd HH_mm_ss"
            archiveNumbering="Date"
            archiveEvery="Year"
            maxArchiveFiles="10"
            
            layout="${longdate} ${uppercase:${level}} ${message}" />
```

Looking forward to v4.5 - seems a lot of stuff will change.

**Comment by nzain** [comment]
Would be nice to include this in the samples - I think the "multiple app instances with separate log files" is a relevant scenario not covered, yet.

**Comment by 304NotModified** [comment]
> Would be nice to include this in the samples -

You could edit the wiki? ;)

**Comment by nzain** [comment]
I selected the [File-target](https://github.com/NLog/NLog/wiki/File-target) wiki page for this addition.
Hopefully, someone finds it useful!

I did not modifiy [Tutorial](https://github.com/NLog/NLog/wiki/Tutorial) (made for beginners) and [Examples](https://github.com/NLog/NLog/wiki/Examples) (I've never looked into that one until today).

**Comment by 304NotModified** [comment]
:+1: thanks!

**Comment by snakefoot** [comment]
@304NotModified This issue has been mostly resolved with NLog 4.5 that allows you to do this:

`<target name="logfile" type="File" fileName="Log-${shortdate}.txt" maxArchiveFiles="7" />`

And it will correctly generate a file-pattern from `fileName=` and cleanup the oldest matching files.

It would make sense to close this multi-feature-request. Since the remaining issue about removing files older than X days already exists here here #311 (Now marked as **up-for-graps**),

**Comment by 304NotModified** [comment]
Agree on that, thanks!
